Bio/SeqIO/Interfaces.py

```diff
@@ -213,30 +207,11 @@ def clean(self, text: str) -> str:
         """Use this to avoid getting newlines in the output."""
         return text.replace("\n", " ").replace("\r", " ")
 
-    def write_header(self):
-        """Write the file header to the output file."""
-        ##################################################
-        # You MUST implement this method in the subclass #
-        # if the file format defines a file header.      #
-        ##################################################
-
-    def write_footer(self):
-        """Write the file footer to the output file."""
-        ##################################################
-        # You MUST implement this method in the subclass #
-        # if the file format defines a file footer.      #
-        ##################################################
-
```

Comment:
Why are you removing these from the base class? Things like the SFF writer use them.

Reply:
Because they are not needed in the base class, if they are implemented in the SFF writer.

Reply:
The idea of the base class was a generic framework recognizing that some formats have headers and footers, and standardizing how the writers should deal with them. I thought there was a good chance of end users calling the writer objects directly, rather than just via ``SeqIO``.

As it happens, almost all the header and footer free formats are now implemented via https://github.com/biopython/biopython/blob/biopython-184/Bio/SeqIO/__init__.py#L456 leading their original writer classes obsolete. See #1223.

We should probably deprecate those now, and ultimately remove them. It will need minor changes to the test suite to get a list of all supported output file formats.

If we do drop all those simple writers, what is left using the base class will either be a binary format, or need a header and/or footer.

Reply:
How will the base class know which SeqRecord-to-string function to use?

Reply:
We may be misunderstanding each other. The SeqRecord to string functionality is used via ``SeqIO.write()`` and the ``.format()`` method using this dict:

https://github.com/biopython/biopython/blob/biopython-184/Bio/SeqIO/__init__.py#L456

Such formats don't need a writer class at all (they still have one for historical reasons, but those writer classes are obsolete and could be deprecated and removed).

Reply:
I see. But then `SeqIO/__init__.py` has to know which format has a writer class, which format has a SeqRecord-to-string function, and what the class name / function name is for each format. This is now maintained in `_FormatToString`, `_FormatToWriter`. But that is not a clean design because it breaks encapsulation. Also, it requires figuring stuff out at run-time.

A cleaner approach is to have a class method `_format_record` that takes a SeqRecord and returns the appropriately formatted string, and let both `format` and `SequenceWriter` call `_format_record`.

Reply:
Chicken and egg - then the SeqRecord ``.format()`` still needs to know which write class to call.

We have to have a central registry of file formats to parsers/writers somewhere don't we?

Reply:
> Chicken and egg - then the SeqRecord .format() still needs to know which write class to call.
> We have to have a central registry of file formats to parsers/writers somewhere don't we?

We don't.
In Bio.Align, we solve this by loading the appropriate module based on the requested `fmt`, using `AlignmentWriter` in  this module, and then calling the `format_alignment` method of this class.
This scheme has the added benefit that submodules are imported only when needed.
Currently, this is implemented in Bio.Align by creating an instance of `AlignmentWriter`, but we could define `format_alignment` as a class method so we even won't need to create an instance.

Reply:
Do you mean the format name is mapped to a Python module (or perhaps function or class) name? And if the import fails, it is an unsupported format. Clever, but perhaps rather disruptive to add to SeqIO/AlignIO/SearchIO retrospectively.

Reply:
The format name maps to a module.

I don't think we need to add it to AlignIO and SearchIO. But SeqIO is too important to leave it in its current state.

Reply:
This approach encodes the mapping from format name to code in the module filenames, which means a lot of upheaval in the (less used but still) public API.

Also, we'd need a mechanism for subformats, e.g. "fasta-2line". Perhaps the first word maps to the submodule name, and then the full name is used to determine the reader/writer class name. Or you just map this to yet another module name using underscores for hyphens?

Reply:
> This approach encodes the mapping from format name to code in the module filenames, which means a lot of upheaval in the (less used but still) public API.

It doesn't have to be this way. There could be one dictionary that maps format names to module filenames.

> Also, we'd need a mechanism for subformats, e.g. "fasta-2line". Perhaps the first word maps to the submodule name, and then the full name is used to determine the reader/writer class name.
> Or you just map this to yet another module name using underscores for hyphens?

Yes.

Reply:
So is the advantage to you one central dict (format to module), over multiple (format to writer, format to string, format to parser)?

Reply:
That is one advantage, but the bigger one is to have proper encapsulation for each format. In its current form, `__init__.py` must know for each format how it is implemented. That should not be the case.